File: ai-service/ai-engine/api/app/services/content_structure.py
"""Content flow checks for generated and saved newsletter issues."""

from __future__ import annotations

import re
from copy import deepcopy
from dataclasses import dataclass
from typing import Any, Dict, List, Literal, Optional

FenceMatch = re.compile(r"```([a-zA-Z0-9_+-]*)\s*\n([\s\S]*?)```", re.MULTILINE)
SegmentKind = Literal["prose", "code", "diagram", "table"]
Severity = Literal["error", "warning", "info"]


@dataclass(frozen=True)
class ContentCheck:
    id: str
    severity: Severity
    message: str
    suggestion: str = ""
    block_index: Optional[int] = None

    def as_dict(self) -> Dict[str, Any]:
        payload: Dict[str, Any] = {
            "id": self.id,
            "severity": self.severity,
            "message": self.message,
        }
        if self.suggestion:
            payload["suggestion"] = self.suggestion
        if self.block_index is not None:
            payload["block_index"] = self.block_index
        return payload


def _classify_fence(lang: str) -> SegmentKind:
    normalized = (lang or "").strip().lower()
    if normalized in {"mermaid", "d2"}:
        return "diagram"
    return "code"


def _is_table_block(text: str) -> bool:
    lines = [line.strip() for line in text.splitlines() if line.strip()]
    if len(lines) < 2:
        return False
    return all(line.startswith("|") and "|" in line[1:] for line in lines)


def _parse_block_segments(text: str, block_index: int) -> List[Dict[str, Any]]:
    source = (text or "").replace("\r\n", "\n")
    if not source.strip():
        return []

    segments: List[Dict[str, Any]] = []
    last_index = 0
    for match in FenceMatch.finditer(source):
        prose = source[last_index : match.start()].strip()
        if prose:
            segments.append(
                {
                    "kind": "table" if _is_table_block(prose) else "prose",
                    "lang": "",
                    "prose_chars": len(prose),
                    "block_index": block_index,
                }
            )
        lang = match.group(1) or ""
        segments.append(
            {
                "kind": _classify_fence(lang),
                "lang": lang,
                "prose_chars": 0,
                "block_index": block_index,
            }
        )
        last_index = match.end()

    tail = source[last_index:].strip()
    if tail:
        segments.append(
            {
                "kind": "table" if _is_table_block(tail) else "prose",
                "lang": "",
                "prose_chars": len(tail),
                "block_index": block_index,
            }
        )
    if not segments and source.strip():
        segments.append(
            {"kind": "prose", "lang": "", "prose_chars": len(source.strip()), "block_index": block_index}
        )
    return segments


def _is_technical(kind: SegmentKind) -> bool:
    return kind in {"code", "diagram"}


def _push(checks: List[ContentCheck], seen: set[str], check: ContentCheck) -> None:
    key = f"{check.id}:{check.block_index if check.block_index is not None else 'all'}"
    if key in seen:
        return
    seen.add(key)
    checks.append(check)


def analyze_content_structure(issue: Dict[str, Any]) -> List[ContentCheck]:
    blocks = [block for block in (issue.get("content_blocks") or []) if isinstance(block, dict)]
    visuals = [
        spec
        for spec in (issue.get("visual_specs") or [])
        if isinstance(spec, dict) and str(spec.get("content") or "").strip()
    ]
    checks: List[ContentCheck] = []
    seen: set[str] = set()
    all_segments: List[Dict[str, Any]] = []

    for block_index, block in enumerate(blocks):
        segments = _parse_block_segments(str(block.get("text") or ""), block_index)
        all_segments.extend(segments)
        if not segments:
            continue

        first = segments[0]
        last = segments[-1]
        if _is_technical(first["kind"]):
            label = "diagram" if first["kind"] == "diagram" else "code block"
            _push(
                checks,
                seen,
                ContentCheck(
                    id="block_opens_with_technical",
                    severity="warning",
                    message=f"Block {block_index + 1} opens with a {label}.",
                    suggestion="Add a short intro sentence before diagrams and code so readers know why they matter.",
                    block_index=block_index,
                ),
            )
        if len(segments) > 1 and _is_technical(last["kind"]):
            label = "diagram" if last["kind"] == "diagram" else "code block"
            _push(
                checks,
                seen,
                ContentCheck(
                    id="block_ends_with_technical",
                    severity="warning",
                    message=f"Block {block_index + 1} ends with a {label}.",
                    suggestion="Follow technical sections with a takeaway sentence or transition.",
                    block_index=block_index,
                ),
            )

        for index in range(len(segments) - 1):
            current = segments[index]
            nxt = segments[index + 1]
            if current["kind"] == "diagram" and nxt["kind"] == "diagram":
                _push(
                    checks,
                    seen,
                    ContentCheck(
                        id="consecutive_diagrams_in_block",
                        severity="error",
                        message=f"Block {block_index + 1} places two diagrams back-to-back.",
                        suggestion="Separate diagrams with prose that explains each step before showing the next one.",
                        block_index=block_index,
                    ),
                )
            if current["kind"] == "code" and nxt["kind"] == "code":
                _push(
                    checks,
                    seen,
                    ContentCheck(
                        id="consecutive_code_in_block",
                        severity="warning",
                        message=f"Block {block_index + 1} stacks code blocks without explanation between them.",
                        suggestion="Explain what the first snippet does before showing another one.",
                        block_index=block_index,
                    ),
                )
            if nxt["kind"] == "diagram" and current["kind"] == "prose" and current["prose_chars"] < 48:
                _push(
                    checks,
                    seen,
                    ContentCheck(
                        id="thin_context_before_diagram",
                        severity="warning",
                        message=f"Block {block_index + 1} has very little setup before a diagram.",
                        suggestion="Add a sentence or two describing what the reader should look for in the diagram.",
                        block_index=block_index,
                    ),
                )

        for segment in segments:
            if segment["kind"] == "code" and not str(segment.get("lang") or "").strip():
                _push(
                    checks,
                    seen,
                    ContentCheck(
                        id="code_missing_language",
                        severity="info",
                        message=f"Block {block_index + 1} has a code fence without a language tag.",
                        suggestion="Use ```bash, ```python, ```yaml, etc. so email clients render it cleanly.",
                        block_index=block_index,
                    ),
                )

    if all_segments:
        if _is_technical(all_segments[0]["kind"]):
            _push(
                checks,
                seen,
                ContentCheck(
                    id="email_opens_with_technical",
                    severity="error",
                    message="The email opens with a diagram or code block before any introduction.",
                    suggestion="Start with a hook paragraph, then introduce visuals and snippets.",
                ),
            )
        if _is_technical(all_segments[-1]["kind"]):
            _push(
                checks,
                seen,
                ContentCheck(
                    id="email_ends_with_technical",
                    severity="warning",
                    message="The email ends on a diagram or code block.",
                    suggestion="Close with a summary, next step, or call to action after the last technical element.",
                ),
            )

        diagram_streak = 0
        code_streak = 0
        for segment in all_segments:
            kind = segment["kind"]
            if kind == "diagram":
                diagram_streak += 1
                code_streak = 0
            elif kind == "code":
                code_streak += 1
                diagram_streak = 0
            else:
                diagram_streak = 0
                code_streak = 0
            if diagram_streak >= 2:
                _push(
                    checks,
                    seen,
                    ContentCheck(
                        id="consecutive_diagrams",
                        severity="error",
                        message="Multiple diagrams appear in a row across the email.",
                        suggestion="Alternate diagrams with explanatory prose so each one lands clearly.",
                        block_index=segment["block_index"],
                    ),
                )
                diagram_streak = 0
            if code_streak >= 3:
                _push(
                    checks,
                    seen,
                    ContentCheck(
                        id="consecutive_code_blocks",
                        severity="warning",
                        message="Three or more code blocks appear back-to-back.",
                        suggestion="Break up long code sequences with commentary or partial snippets.",
                        block_index=segment["block_index"],
                    ),
                )
                code_streak = 0

    if len(blocks) > 1:
        for block_index, block in enumerate(blocks):
            text = str(block.get("text") or "")
            if "```" in text and not str(block.get("title") or "").strip():
                _push(
                    checks,
                    seen,
                    ContentCheck(
                        id="missing_block_title",
                        severity="info",
                        message=f"Block {block_index + 1} includes code or diagrams but has no title.",
                        suggestion="Add a block title so readers know what each section covers.",
                        block_index=block_index,
                    ),
                )

    inline_diagrams = any(segment["kind"] == "diagram" for segment in all_segments)
    if visuals and inline_diagrams:
        _push(
            checks,
            seen,
            ContentCheck(
                id="mixed_diagram_placement",
                severity="info",
                message="This issue uses both inline diagrams and detached visuals.",
                suggestion="Keep inline diagrams near the text they explain; reserve detached visuals for summary figures at the end.",
            ),
        )

    if visuals and blocks:
        last_index = len(blocks) - 1
        last_segments = _parse_block_segments(str(blocks[last_index].get("text") or ""), last_index)
        if last_segments and _is_technical(last_segments[-1]["kind"]):
            _push(
                checks,
                seen,
                ContentCheck(
                    id="detached_visual_after_technical_end",
                    severity="warning",
                    message="Detached visuals will render after a body that already ends on code or a diagram.",
                    suggestion="Add a short closing paragraph before detached visuals, or move the diagram inline.",
                    block_index=last_index,
                ),
            )

    return checks


def structure_issue_errors(checks: List[ContentCheck]) -> List[ContentCheck]:
    return [check for check in checks if check.severity == "error"]


def normalize_issue_content_order(issue: Dict[str, Any]) -> Dict[str, Any]:
    """Apply safe ordering fixes before persisting generated content."""
    normalized = deepcopy(issue)
    blocks = normalized.get("content_blocks") or []
    updated_blocks = []
    for block in blocks:
        if not isinstance(block, dict):
            updated_blocks.append(block)
            continue
        text = str(block.get("text") or "").lstrip()
        title = str(block.get("title") or "").strip()
        if title and text.startswith("```") and not text.startswith(f"## {title}"):
            text = f"## {title}\n\n{text}"
        updated_blocks.append({**block, "text": text})
    normalized["content_blocks"] = updated_blocks
    return normalized


def structure_feedback(checks: List[ContentCheck]) -> str:
    if not checks:
        return ""
    lines = ["Fix content flow issues before finalizing:"]
    for check in checks[:6]:
        lines.append(f"- {check.message} {check.suggestion}".strip())
    return "\n".join(lines)

File: frontend/web/src/lib/contentStructure.ts
export type ContentCheckSeverity = 'error' | 'warning' | 'info';

export type ContentCheck = {
  id: string;
  severity: ContentCheckSeverity;
  message: string;
  suggestion?: string;
  blockIndex?: number;
};

export type ContentStructureInput = {
  content_blocks?: Array<{
    title?: string;
    text?: string;
  }>;
  visual_specs?: Array<{
    type?: string;
    content?: string;
    alt_text?: string;
  }>;
};

type SegmentKind = 'prose' | 'code' | 'diagram' | 'table';

type Segment = {
  kind: SegmentKind;
  lang: string;
  proseChars: number;
  blockIndex: number;
};

const FENCE_RE = /```([a-zA-Z0-9_+-]*)\s*\n([\s\S]*?)```/g;

function isTableBlock(text: string) {
  const lines = text.split('\n').filter((line) => line.trim());
  if (lines.length < 2) return false;
  return lines.every((line) => line.trim().startsWith('|') && line.includes('|', 1));
}

function classifyFence(lang: string): SegmentKind {
  const normalized = lang.toLowerCase();
  if (normalized === 'mermaid' || normalized === 'd2') return 'diagram';
  return 'code';
}

function parseBlockSegments(text: string, blockIndex: number): Segment[] {
  const source = (text || '').replace(/\r\n/g, '\n');
  if (!source.trim()) return [];

  const segments: Segment[] = [];
  let lastIndex = 0;
  let match: RegExpExecArray | null;
  FENCE_RE.lastIndex = 0;

  while ((match = FENCE_RE.exec(source)) !== null) {
    const prose = source.slice(lastIndex, match.index);
    const proseTrimmed = prose.trim();
    if (proseTrimmed) {
      segments.push({
        kind: isTableBlock(proseTrimmed) ? 'table' : 'prose',
        lang: '',
        proseChars: proseTrimmed.length,
        blockIndex,
      });
    }
    segments.push({
      kind: classifyFence(match[1] || ''),
      lang: match[1] || '',
      proseChars: 0,
      blockIndex,
    });
    lastIndex = match.index + match[0].length;
  }

  const tail = source.slice(lastIndex).trim();
  if (tail) {
    segments.push({
      kind: isTableBlock(tail) ? 'table' : 'prose',
      lang: '',
      proseChars: tail.length,
      blockIndex,
    });
  }

  if (segments.length === 0 && source.trim()) {
    segments.push({
      kind: 'prose',
      lang: '',
      proseChars: source.trim().length,
      blockIndex,
    });
  }

  return segments;
}

function isTechnical(kind: SegmentKind) {
  return kind === 'code' || kind === 'diagram';
}

function pushCheck(
  checks: ContentCheck[],
  seen: Set<string>,
  check: ContentCheck,
) {
  const key = `${check.id}:${check.blockIndex ?? 'all'}`;
  if (seen.has(key)) return;
  seen.add(key);
  checks.push(check);
}

export function analyzeContentStructure(input: ContentStructureInput): ContentCheck[] {
  const blocks = input.content_blocks ?? [];
  const visuals = (input.visual_specs ?? []).filter((visual) => (visual.content || '').trim());
  const checks: ContentCheck[] = [];
  const seen = new Set<string>();

  const allSegments: Segment[] = [];
  blocks.forEach((block, blockIndex) => {
    const segments = parseBlockSegments(block.text || '', blockIndex);
    allSegments.push(...segments);

    if (segments.length === 0) return;

    const first = segments[0];
    const last = segments[segments.length - 1];

    if (isTechnical(first.kind)) {
      pushCheck(checks, seen, {
        id: 'block_opens_with_technical',
        severity: 'warning',
        message: `Block ${blockIndex + 1} opens with a ${first.kind === 'diagram' ? 'diagram' : 'code block'}.`,
        suggestion: 'Add a short intro sentence before diagrams and code so readers know why they matter.',
        blockIndex,
      });
    }

    if (segments.length > 1 && isTechnical(last.kind)) {
      pushCheck(checks, seen, {
        id: 'block_ends_with_technical',
        severity: 'warning',
        message: `Block ${blockIndex + 1} ends with a ${last.kind === 'diagram' ? 'diagram' : 'code block'}.`,
        suggestion: 'Follow technical sections with a takeaway sentence or transition.',
        blockIndex,
      });
    }

    for (let i = 0; i < segments.length - 1; i += 1) {
      const current = segments[i];
      const next = segments[i + 1];
      if (current.kind === 'diagram' && next.kind === 'diagram') {
        pushCheck(checks, seen, {
          id: 'consecutive_diagrams_in_block',
          severity: 'error',
          message: `Block ${blockIndex + 1} places two diagrams back-to-back.`,
          suggestion: 'Separate diagrams with prose that explains each step before showing the next one.',
          blockIndex,
        });
      }
      if (current.kind === 'code' && next.kind === 'code') {
        pushCheck(checks, seen, {
          id: 'consecutive_code_in_block',
          severity: 'warning',
          message: `Block ${blockIndex + 1} stacks code blocks without explanation between them.`,
          suggestion: 'Explain what the first snippet does before showing another one.',
          blockIndex,
        });
      }
      if (next.kind === 'diagram' && current.kind === 'prose' && current.proseChars < 48) {
        pushCheck(checks, seen, {
          id: 'thin_context_before_diagram',
          severity: 'warning',
          message: `Block ${blockIndex + 1} has very little setup before a diagram.`,
          suggestion: 'Add a sentence or two describing what the reader should look for in the diagram.',
          blockIndex,
        });
      }
      if (next.kind === 'code' && current.kind === 'prose' && current.proseChars < 32) {
        pushCheck(checks, seen, {
          id: 'thin_context_before_code',
          severity: 'info',
          message: `Block ${blockIndex + 1} jumps into code with minimal setup.`,
          suggestion: 'Briefly say what the snippet demonstrates before the fence.',
          blockIndex,
        });
      }
    }

    segments.forEach((segment) => {
      if (segment.kind === 'code' && !segment.lang.trim()) {
        pushCheck(checks, seen, {
          id: 'code_missing_language',
          severity: 'info',
          message: `Block ${blockIndex + 1} has a code fence without a language tag.`,
          suggestion: 'Use ```bash, ```python, ```yaml, etc. so email clients render it cleanly.',
          blockIndex,
        });
      }
    });
  });

  if (allSegments.length > 0) {
    const firstOverall = allSegments[0];
    const lastOverall = allSegments[allSegments.length - 1];

    if (isTechnical(firstOverall.kind)) {
      pushCheck(checks, seen, {
        id: 'email_opens_with_technical',
        severity: 'error',
        message: 'The email opens with a diagram or code block before any introduction.',
        suggestion: 'Start with a hook paragraph, then introduce visuals and snippets.',
      });
    }

    if (isTechnical(lastOverall.kind)) {
      pushCheck(checks, seen, {
        id: 'email_ends_with_technical',
        severity: 'warning',
        message: 'The email ends on a diagram or code block.',
        suggestion: 'Close with a summary, next step, or call to action after the last technical element.',
      });
    }

    let diagramStreak = 0;
    let codeStreak = 0;
    allSegments.forEach((segment) => {
      if (segment.kind === 'diagram') {
        diagramStreak += 1;
        codeStreak = 0;
      } else if (segment.kind === 'code') {
        codeStreak += 1;
        diagramStreak = 0;
      } else if (segment.kind === 'prose') {
        diagramStreak = 0;
        codeStreak = 0;
      }
      if (diagramStreak >= 2) {
        pushCheck(checks, seen, {
          id: 'consecutive_diagrams',
          severity: 'error',
          message: 'Multiple diagrams appear in a row across the email.',
          suggestion: 'Alternate diagrams with explanatory prose so each one lands clearly.',
          blockIndex: segment.blockIndex,
        });
        diagramStreak = 0;
      }
      if (codeStreak >= 3) {
        pushCheck(checks, seen, {
          id: 'consecutive_code_blocks',
          severity: 'warning',
          message: 'Three or more code blocks appear back-to-back.',
          suggestion: 'Break up long code sequences with commentary or partial snippets.',
          blockIndex: segment.blockIndex,
        });
        codeStreak = 0;
      }
    });
  }

  if (blocks.length > 1) {
    blocks.forEach((block, blockIndex) => {
      const hasTechnical = /```/.test(block.text || '');
      if (hasTechnical && !block.title?.trim()) {
        pushCheck(checks, seen, {
          id: 'missing_block_title',
          severity: 'info',
          message: `Block ${blockIndex + 1} includes code or diagrams but has no title.`,
          suggestion: 'Add a block title so readers know what each section covers.',
          blockIndex,
        });
      }
    });
  }

  const inlineDiagrams = allSegments.some((segment) => segment.kind === 'diagram');
  if (visuals.length > 0 && inlineDiagrams) {
    pushCheck(checks, seen, {
      id: 'mixed_diagram_placement',
      severity: 'info',
      message: 'This issue uses both inline diagrams and detached visuals.',
      suggestion: 'Keep inline diagrams near the text they explain; reserve detached visuals for summary figures at the end.',
    });
  }

  if (visuals.length > 0 && blocks.length > 0) {
    const lastBlock = blocks[blocks.length - 1];
    const lastSegments = parseBlockSegments(lastBlock.text || '', blocks.length - 1);
    const lastKind = lastSegments[lastSegments.length - 1]?.kind;
    if (lastKind && isTechnical(lastKind)) {
      pushCheck(checks, seen, {
        id: 'detached_visual_after_technical_end',
        severity: 'warning',
        message: 'Detached visuals will render after a body that already ends on code or a diagram.',
        suggestion: 'Add a short closing paragraph before detached visuals, or move the diagram inline.',
        blockIndex: blocks.length - 1,
      });
    }
  }

  return checks;
}

export function contentStructureScore(checks: ContentCheck[]) {
  const errors = checks.filter((check) => check.severity === 'error').length;
  const warnings = checks.filter((check) => check.severity === 'warning').length;
  const infos = checks.filter((check) => check.severity === 'info').length;
  const score = Math.max(0, 100 - errors * 20 - warnings * 8 - infos * 2);
  return { score, errors, warnings, infos };
}

export function summarizeContentChecks(checks: ContentCheck[]) {
  const { score, errors, warnings } = contentStructureScore(checks);
  if (errors > 0) return { label: 'Needs reordering', tone: 'error' as const, score };
  if (warnings > 0) return { label: 'Could be tighter', tone: 'warning' as const, score };
  if (checks.length > 0) return { label: 'Good flow', tone: 'info' as const, score };
  return { label: 'Well structured', tone: 'success' as const, score };
}

export function contentStructureFromIssue(contentJson?: string | Record<string, unknown> | null): ContentStructureInput {
  let parsed: Record<string, unknown> | null = null;
  if (contentJson && typeof contentJson === 'object') {
    parsed = contentJson as Record<string, unknown>;
  } else if (typeof contentJson === 'string' && contentJson) {
    try {
      parsed = JSON.parse(contentJson) as Record<string, unknown>;
    } catch {
      parsed = null;
    }
  }
  return {
    content_blocks: Array.isArray(parsed?.content_blocks)
      ? (parsed.content_blocks as ContentStructureInput['content_blocks'])
      : [],
    visual_specs: Array.isArray(parsed?.visual_specs)
      ? (parsed.visual_specs as ContentStructureInput['visual_specs'])
      : [],
  };
}
